File: examples/python/indexer_feegrant_v1_query_messages.py
#!/usr/bin/env python3
"""
xion.indexer.feegrant.v1 Query Messages — Protobuf Examples

This file demonstrates how to construct, encode, and decode
query messages from the Feegrant Indexer Module (xion.indexer.feegrant.v1).

Queries covered:
- QueryAllowance: Get a specific fee allowance
- QueryAllowances: List all allowances for a grantee
- QueryAllowancesByGranter: List all allowances issued by a granter

The Feegrant Indexer provides optimized queries for fee allowances,
enabling efficient lookup of who pays fees for whom.

Note: This is an indexer module with query-only messages.
To create/revoke fee grants, use the standard cosmos.feegrant.v1beta1 module.
"""

import sys
import os

# Add the python types to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python/types'))

from xion.indexer.feegrant.v1 import query_pb2 as feegrant_query
from google.protobuf.json_format import MessageToJson


# =============================================================================
# QueryAllowanceRequest / Response
# =============================================================================
# Get a specific fee allowance granted from granter to grantee.
# Endpoint: GET /xion/indexer/feegrant/v1/allowance/{granter}/{grantee}
# Type URL: /xion.indexer.feegrant.v1.QueryAllowanceRequest

def query_allowance_examples():
    print('=== QueryAllowance Examples ===\n')

    # 1. Construct request
    request = feegrant_query.QueryAllowanceRequest()
    request.granter = 'xion1granter...'  # Who is paying fees
    request.grantee = 'xion1grantee...'  # Who benefits from fee payment

    print('1. Constructed QueryAllowanceRequest:')
    print(f'   granter: {request.granter}')
    print(f'   grantee: {request.grantee}')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowanceRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # 3. Decode request
    decoded_request = feegrant_query.QueryAllowanceRequest()
    decoded_request.ParseFromString(encoded_request)
    print(f'\n3. Decoded granter: {decoded_request.granter}')
    print(f'   Decoded grantee: {decoded_request.grantee}')

    # === Response ===
    print('\n--- Response ---')

    # Note: Response contains cosmos.feegrant.v1beta1.Grant
    response = feegrant_query.QueryAllowanceResponse()
    # response.allowance contains:
    #   - granter: string
    #   - grantee: string
    #   - allowance: Any (BasicAllowance, PeriodicAllowance, AllowedMsgAllowance)

    print('4. Response structure:')
    print('   allowance: cosmos.feegrant.v1beta1.Grant')
    print('     - granter: string')
    print('     - grantee: string')
    print('     - allowance: Any (BasicAllowance, PeriodicAllowance, etc.)')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowanceResponse')

    # 4. JSON format
    json_str = MessageToJson(request)
    print(f'\n5. Request JSON format:')
    print(f'   {json_str}')

    return request, response


# =============================================================================
# QueryAllowancesRequest / Response
# =============================================================================
# List all fee allowances for a specific grantee (who can have fees paid for them).
# Endpoint: GET /xion/indexer/feegrant/v1/allowances/{grantee}
# Type URL: /xion.indexer.feegrant.v1.QueryAllowancesRequest

def query_allowances_examples():
    print('\n=== QueryAllowances Examples ===\n')

    # 1. Construct request
    request = feegrant_query.QueryAllowancesRequest()
    request.grantee = 'xion1grantee...'  # List all grants TO this address
    request.pagination.limit = 20
    request.pagination.count_total = True

    print('1. Constructed QueryAllowancesRequest:')
    print(f'   grantee: {request.grantee}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowancesRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = feegrant_query.QueryAllowancesResponse()

    print('3. Response structure:')
    print('   allowances: [cosmos.feegrant.v1beta1.Grant]')
    print('   pagination: PageResponse')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowancesResponse')

    return request, response


# =============================================================================
# QueryAllowancesByGranterRequest / Response
# =============================================================================
# List all fee allowances issued by a specific granter (who pays fees for others).
# Endpoint: GET /xion/indexer/feegrant/v1/issued/{granter}
# Type URL: /xion.indexer.feegrant.v1.QueryAllowancesByGranterRequest

def query_allowances_by_granter_examples():
    print('\n=== QueryAllowancesByGranter Examples ===\n')

    # 1. Construct request
    request = feegrant_query.QueryAllowancesByGranterRequest()
    request.granter = 'xion1granter...'  # List all grants FROM this address
    request.pagination.limit = 20
    request.pagination.count_total = True

    print('1. Constructed QueryAllowancesByGranterRequest:')
    print(f'   granter: {request.granter}')
    print(f'   pagination.limit: {request.pagination.limit}')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowancesByGranterRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = feegrant_query.QueryAllowancesByGranterResponse()

    print('3. Response structure:')
    print('   allowances: [cosmos.feegrant.v1beta1.Grant]')
    print('   pagination: PageResponse')
    print(f'   typeUrl: /xion.indexer.feegrant.v1.QueryAllowancesByGranterResponse')

    return request, response


# =============================================================================
# Understanding Fee Grants
# =============================================================================

def feegrant_explanation():
    print('\n=== Understanding Fee Grants ===\n')

    print('Feegrant allows one account to pay transaction fees for another.\n')

    print('Key Concepts:\n')
    print('  1. Granter: The account paying fees')
    print('  2. Grantee: The account benefiting (their txs are fee-free)')
    print('  3. Allowance: Rules for fee payment (limits, expiry, etc.)\n')

    print('Allowance Types:\n')
    allowance_types = [
        ('BasicAllowance', 'Simple spend limit with optional expiration'),
        ('PeriodicAllowance', 'Allowance that resets periodically'),
        ('AllowedMsgAllowance', 'Wraps another allowance, restricts msg types'),
    ]

    for allowance_type, desc in allowance_types:
        print(f'  • {allowance_type}:')
        print(f'    {desc}\n')

    print('Use Cases:\n')
    print('  • Onboarding: Pay fees for new users (gasless first tx)')
    print('  • dApps: Pay user fees for better UX')
    print('  • Abstract accounts: Enable meta-transactions')
    print('  • DAOs: Treasury pays member transaction fees\n')

    print('Indexer vs Standard Feegrant:\n')
    print('  • This module provides OPTIMIZED QUERIES for allowances')
    print('  • To CREATE grants, use: cosmos.feegrant.v1beta1.MsgGrantAllowance')
    print('  • To REVOKE grants, use: cosmos.feegrant.v1beta1.MsgRevokeAllowance')


# =============================================================================
# Run All Examples
# =============================================================================

def main():
    print('╔════════════════════════════════════════════════════════════════╗')
    print('║  xion.indexer.feegrant.v1 Query Messages — Protobuf Examples   ║')
    print('╚════════════════════════════════════════════════════════════════╝\n')

    query_allowance_examples()
    query_allowances_examples()
    query_allowances_by_granter_examples()
    feegrant_explanation()

    print('\n═══════════════════════════════════════════════════════════════')
    print('Type URL Reference (Requests):')
    print('  QueryAllowanceRequest:          /xion.indexer.feegrant.v1.QueryAllowanceRequest')
    print('  QueryAllowancesRequest:         /xion.indexer.feegrant.v1.QueryAllowancesRequest')
    print('  QueryAllowancesByGranterRequest:/xion.indexer.feegrant.v1.QueryAllowancesByGranterRequest')
    print('\nType URL Reference (Responses):')
    print('  QueryAllowanceResponse:          /xion.indexer.feegrant.v1.QueryAllowanceResponse')
    print('  QueryAllowancesResponse:         /xion.indexer.feegrant.v1.QueryAllowancesResponse')
    print('  QueryAllowancesByGranterResponse:/xion.indexer.feegrant.v1.QueryAllowancesByGranterResponse')
    print('\nREST Endpoints:')
    print('  GET /xion/indexer/feegrant/v1/allowance/{granter}/{grantee}')
    print('  GET /xion/indexer/feegrant/v1/allowances/{grantee}')
    print('  GET /xion/indexer/feegrant/v1/issued/{granter}')
    print('\nNote: This is an indexer module (query-only).')
    print('Use cosmos.feegrant.v1beta1 for grant/revoke transactions.')
    print('═══════════════════════════════════════════════════════════════')


if __name__ == '__main__':
    main()

File: examples/rust/examples/indexer_feegrant_v1_query_messages.rs
//! xion.indexer.feegrant.v1 Query Messages — Protobuf Examples
//!
//! This file demonstrates how to construct, encode, and decode
//! query messages from the Feegrant Indexer Module (xion.indexer.feegrant.v1).
//!
//! Queries covered:
//! - QueryAllowance: Get specific fee allowance between granter and grantee
//! - QueryAllowances: Get all fee allowances received by a grantee
//! - QueryAllowancesByGranter: Get all fee allowances issued by a granter
//!
//! Note: This is an indexer module - it provides efficient queries for
//! fee allowances that would otherwise require scanning the entire state.
//!
//! **IMPORTANT**: This example requires the `xion_indexer_feegrant_v1` module to be
//! exported by the xion-types crate. If you get an "unresolved import" error,
//! the crate needs to be updated to include the indexer module exports.

use prost::Message;

use xion_types::types::xion_indexer_feegrant_v1::{
    QueryAllowanceRequest, QueryAllowanceResponse, QueryAllowancesByGranterRequest,
    QueryAllowancesByGranterResponse, QueryAllowancesRequest, QueryAllowancesResponse,
};

use xion_types::types::cosmos_base_query_v1beta1::{PageRequest, PageResponse};

// =============================================================================
// QueryAllowance
// =============================================================================
// Get specific fee allowance between granter and grantee.
// Endpoint: GET /xion/indexer/feegrant/v1/allowance/{granter}/{grantee}
// Type URL: /xion.indexer.feegrant.v1.QueryAllowanceRequest

fn query_allowance_examples() {
    println!("=== QueryAllowance Examples ===\n");

    // 1. Construct request
    let request = QueryAllowanceRequest {
        granter: "xion1granter...".to_string(),
        grantee: "xion1grantee...".to_string(),
    };

    println!("1. Constructed QueryAllowanceRequest:");
    println!("   granter: {}", request.granter);
    println!("   grantee: {}", request.grantee);

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryAllowanceRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded:");
    println!("   granter: {}", decoded_request.granter);
    println!("   grantee: {}", decoded_request.grantee);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains allowance: Option<cosmos.feegrant.v1beta1.Grant>
    let response = QueryAllowanceResponse {
        allowance: None, // Would contain Grant object
    };

    println!("4. Response structure:");
    println!("   allowance: Option<cosmos.feegrant.v1beta1.Grant>");
    println!("     - granter: String");
    println!("     - grantee: String");
    println!("     - allowance: Any (BasicAllowance or PeriodicAllowance)");
    println!("   has_allowance: {}", response.allowance.is_some());
}

// =============================================================================
// QueryAllowances
// =============================================================================
// Get all fee allowances received by a grantee.
// Endpoint: GET /xion/indexer/feegrant/v1/allowances/{grantee}
// Type URL: /xion.indexer.feegrant.v1.QueryAllowancesRequest

fn query_allowances_examples() {
    println!("\n=== QueryAllowances Examples ===\n");

    // 1. Construct request
    let request = QueryAllowancesRequest {
        grantee: "xion1grantee...".to_string(),
        pagination: Some(PageRequest {
            key: vec![],
            offset: 0,
            limit: 10,
            count_total: true,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryAllowancesRequest:");
    println!("   grantee: {}", request.grantee);
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
        println!("   pagination.count_total: {}", pagination.count_total);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request = QueryAllowancesRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded grantee: {}", decoded_request.grantee);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains allowances: Vec<cosmos.feegrant.v1beta1.Grant>
    let response = QueryAllowancesResponse {
        allowances: vec![], // Would contain Grant objects
        pagination: Some(PageResponse {
            next_key: vec![],
            total: 2,
        }),
    };

    println!("4. Response structure:");
    println!("   allowances: Vec<cosmos.feegrant.v1beta1.Grant>");
    println!("   pagination: PageResponse");
    println!("   allowances count: {}", response.allowances.len());
    if let Some(ref pagination) = response.pagination {
        println!("   pagination.total: {}", pagination.total);
    }

    // Use case
    println!("\n--- Use Case ---");
    println!("5. Find all fee grants received by an account:");
    println!("   - Check if someone is paying your transaction fees");
    println!("   - List all accounts sponsoring this wallet");
    println!("   - Useful for gasless/meta-transaction flows");
}

// =============================================================================
// QueryAllowancesByGranter
// =============================================================================
// Get all fee allowances issued by a granter.
// Endpoint: GET /xion/indexer/feegrant/v1/issued/{granter}
// Type URL: /xion.indexer.feegrant.v1.QueryAllowancesByGranterRequest

fn query_allowances_by_granter_examples() {
    println!("\n=== QueryAllowancesByGranter Examples ===\n");

    // 1. Construct request
    let request = QueryAllowancesByGranterRequest {
        granter: "xion1granter...".to_string(),
        pagination: Some(PageRequest {
            key: vec![],
            offset: 0,
            limit: 20,
            count_total: true,
            reverse: false,
        }),
    };

    println!("1. Constructed QueryAllowancesByGranterRequest:");
    println!("   granter: {}", request.granter);
    if let Some(ref pagination) = request.pagination {
        println!("   pagination.limit: {}", pagination.limit);
    }

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request =
        QueryAllowancesByGranterRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded granter: {}", decoded_request.granter);

    // === Response ===
    println!("\n--- Response ---");

    // Response contains allowances: Vec<cosmos.feegrant.v1beta1.Grant>
    let response = QueryAllowancesByGranterResponse {
        allowances: vec![], // Would contain Grant objects
        pagination: Some(PageResponse {
            next_key: vec![],
            total: 5,
        }),
    };

    println!("4. Response structure:");
    println!("   allowances: Vec<cosmos.feegrant.v1beta1.Grant>");
    println!("   pagination: PageResponse");
    println!("   allowances count: {}", response.allowances.len());
    if let Some(ref pagination) = response.pagination {
        println!("   pagination.total: {}", pagination.total);
    }

    // Use case
    println!("\n--- Use Case ---");
    println!("5. Find all fee grants issued by an account:");
    println!("   - Audit who you're sponsoring for transaction fees");
    println!("   - Monitor fee grant expenditure");
    println!("   - Manage gasless transaction sponsorship");
}

// =============================================================================
// Understanding Feegrant Indexer
// =============================================================================

fn feegrant_indexer_explanation() {
    println!("\n=== Understanding Feegrant Indexer ===\n");

    println!("The Feegrant Indexer provides efficient queries for fee allowances.\n");

    println!("Key Concepts:\n");

    println!("1. Fee Allowance:");
    println!("   - Allows granter to pay transaction fees for grantee");
    println!("   - Enables gasless transactions for end users");
    println!("   - Can have spending limits and expiration\n");

    println!("2. Allowance Types:");
    let allowance_types = [
        ("BasicAllowance", "Simple spend limit with optional expiration"),
        (
            "PeriodicAllowance",
            "Resets spending limit periodically (daily, weekly, etc.)",
        ),
        (
            "AllowedMsgAllowance",
            "Wraps another allowance with message type restrictions",
        ),
    ];
    for (name, desc) in &allowance_types {
        println!("   - {}: {}", name, desc);
    }

    println!("\n3. Why Use the Indexer?");
    println!("   - Standard feegrant queries require knowing both granter AND grantee");
    println!("   - Indexer allows querying by granter OR grantee alone");
    println!("   - Essential for apps showing all sponsored users\n");

    println!("4. Common Use Cases:");
    println!("   - Gasless onboarding: App pays fees for new users");
    println!("   - Subscription services: Pay fees for premium users");
    println!("   - DAO operations: Treasury pays member transaction fees");
}

// =============================================================================
// Run All Examples
// =============================================================================

fn main() {
    println!("╔════════════════════════════════════════════════════════════╗");
    println!("║  xion.indexer.feegrant.v1 Query Messages — Examples        ║");
    println!("╚════════════════════════════════════════════════════════════╝\n");

    query_allowance_examples();
    query_allowances_examples();
    query_allowances_by_granter_examples();
    feegrant_indexer_explanation();

    println!("\n═══════════════════════════════════════════════════════════════");
    println!("Type URL Reference:");
    println!("  QueryAllowanceRequest:          /xion.indexer.feegrant.v1.QueryAllowanceRequest");
    println!("  QueryAllowancesRequest:         /xion.indexer.feegrant.v1.QueryAllowancesRequest");
    println!(
        "  QueryAllowancesByGranterRequest: /xion.indexer.feegrant.v1.QueryAllowancesByGranterRequest"
    );
    println!("\nREST Endpoints:");
    println!("  GET /xion/indexer/feegrant/v1/allowance/{{granter}}/{{grantee}}");
    println!("  GET /xion/indexer/feegrant/v1/allowances/{{grantee}}");
    println!("  GET /xion/indexer/feegrant/v1/issued/{{granter}}");
    println!("\nNote: This is a query-only indexer module.");
    println!("═══════════════════════════════════════════════════════════════");
}
